### ISSUE TITLE
[DOCUMENTATION] Glyph Line, Text, and Fill defaults in theme yaml are undocumented

### ISSUE THREAD

**Issue by jeremy9959** [bokeh/bokeh#9693]
The [reference manual page for themes](https://docs.bokeh.org/en/latest/docs/reference/themes.html) does not document the ability to set default line, text, and fill attributes for glyphs in the theme.yaml file.  For example this is valid:
```yaml
attrs:
  Axis: {}
  Grid: {}
  Plot: {}
  Title:
    text_color: '#f4a4c0'
  Toolbar: {}
fill_defaults:
  fill_color: '#e32400'
line_defaults:
  line_color: '#2b1276'
text_defaults:
  text_color: '#7a4900'
  text_font: Calibri
```

and has the effect, for example, of setting the default font in all Text Glyphs to "Calibri" unless specifically overridden.  The relevant code is in ```theme.py``` where the constructor for the ```Theme``` class includes:
```python
    self._line_defaults = self._json.get('line_defaults', _empty_dict)
    self._fill_defaults = self._json.get('fill_defaults', _empty_dict)
    self._text_defaults = self._json.get('text_defaults', _empty_dict)
```
and there is the method:
```python
def _add_glyph_defaults(self, cls, props):
        from ..models.glyphs import Glyph
        if issubclass(cls, Glyph):
            if hasattr(cls, "line_alpha"):
                props.update(self._line_defaults)
            if hasattr(cls, "fill_alpha"):
                props.update(self._fill_defaults)
            if hasattr(cls, "text_alpha"):
                props.update(self._text_defaults)
```

The documentation also includes the somewhat cryptic annotation:
```
    The plotting API's defaults override some theme properties. Namely:
    `fill_alpha`, `fill_color`, `line_alpha`, `line_color`, `text_alpha` and
    `text_color`. Those properties should therefore be set explicitly when
    using the plotting API.
```
which merits more explanation.

**Comment by bryevdv** [comment]
Regarding the cryptic comment, there is a long discussion in https://github.com/bokeh/bokeh/issues/9173

TLDR; The plotting API currently computes its own defaults for some values, and *sets them*. Currently actual set values always overrides theme values, so an extension of theming that the Plotting API could plug into would be necessary to make this situation better.

**Comment by bryevdv** [comment]
The more I think about this, the more I would like to get rid of `line_defaults`, etc. We could remove them for 2.0 or deprecated them. If #9339 is implemented then the generic name "line_defaults" is confusing because it really only applies to glyphs... thoughts? 

I am not opposed to having this feature, but the proper name would be more specific: `glyph_line_defaults`, etc. so that it's clear what the settings apply to.

**Comment by jeremy9959** [comment]
I certainly have no attachment to it.  I like the idea of a "Glyph" section or more generally
a section that has

Class:
   attribute: value
...

Where the specification for a class applies to that class and its superclasses unless overridden at a higher level.

**Comment by bryevdv** [comment]
```
Class:
attribute: value
``` 
This is exactly what the current content under `Attrs` is, unless I am mistaken about what you are suggesting?

**Comment by jeremy9959** [comment]
I didn't realize that a specification also applies to superclasses.  Can you say:
```yaml
attrs:
    Glyph:
        text_font: Calibri
```
and it applies to all glyphs?

**Comment by bryevdv** [comment]
Do you mean subclasses? I am not sure we can support that, except via special case (what I was proposing above)

**Comment by jeremy9959** [comment]
I am being dumb. The reason for line_defaults etc is that not all glyphs have line attributes.  So you need a way to set all line colors without enumerating all the types that have lines.

**Comment by jeremy9959** [comment]
Right now if you put
```yaml
attrs:
    Glyph:
        text_font_size: 16pt
```
and then try to use a Line glyph you get
```
AttributeError: 'Line' object has no attribute 'text_font_size'
```

One very naive approach would be to ignore attributes set in the theme file if
the model in question doesn't have that attribute.  Then you could take all of the line-defaults, fill_defaults, and text_defaults and just put them under Glyph
and they'd apply where they exist.

**Comment by bryevdv** [comment]
> One very naive approach would be to ignore attributes set in the theme file if
the model in question doesn't have that attribute. 

Basically I think this is a great idea for all glyphs (specifically, because they all collectively sort-of share the same visual properties interface), but *not* a good idea for not-Glyphs (because in general it more likely indicates a typo, or error, and that should be reported out). 

The current `line_properties` is exactly that, a specific case for glyphs. I think whatever we do we should keep this special case for more "forgiving" property setting confined to glyphs, just better document where and how it is used. Along those lines I was thinking maybe `glyphs` could be an  entirely separate top-level construct. That would make it clear that it is "different" (has forgiving semantics for setting undefined properties)

```
attrs:
    ....

glyphs:
    text_font_size: 16pt
```

We could also make a special case for `Glyph` under `attrs` as you have, I just think that's a little harder to document cleanly.

**Comment by jeremy9959** [comment]
Just to air all options, here is one more: Create a package of "glyph attributes" that includes essentially all of the line, fill, and text options, and give them to every glyph *even if they aren't used by that particular glyph* -- perhaps by making them attributes of the Glpyh class.  Then you don't need to treat glyphs separately, you can specify options for even the unused attributes without raising an exception.

**Comment by bryevdv** [comment]
That would add e.g. add all text properties to all glyphs in the docs, almost none of which can use text properties. That seems to be an invitation for user-confusion and support issues so I am definitely -1 on that.

**Comment by jeremy9959** [comment]
FWIW from the ```Marker``` class documentation:
```
Markers are all subclasses of ``Glyph``. Additionally, they all share the
same common interface providing fill and line properties provided by their
base class ``Marker``. Note that a few glyphs, ``Cross`` and ``X``, only
draw lines. For these the fill property values are ignored. Also note that
the ``Circle`` glyph has some additional properties such as ``radius`` that
other markers do not.
```
There are already ```XYGlpyhs```  and ```ConnectedXYGlyph``` so perhaps there should be subclasses of Glyphs  for which the corresponding groups of attributes apply.

My thought here is that I'm trying to avoid elaborate parsing of yaml theme files.

**Comment by jeremy9959** [comment]
I would want to avoid elaborate parsing of theme yaml files, that's why I like subclassing glyphs so that the subclasses have the right property sets.

**Comment by bryevdv** [comment]
@jeremy9959 I don't really understand what you are suggesting. What you describe is exactly how things already are, with the single exception of Markers. The reason for markers is that do really need to be used completely interchangeably within code in various places (both in our library code and also, as I have seen, often in user code as well.

**Comment by jeremy9959** [comment]
Some glyphs have line properties; some have fill properties, some have text properties. My suggestion is to create subclasses of Glyph characterized by which subset of these the subclass supports.   This sort of exists in Markers and XYGlyphs and ConnectedGlyphs but not really.

**Comment by bryevdv** [comment]
If you mean create purely abstract `FilledGlyph` etc, that seems reasonable though we should add tests to make sure e.g. that all `FilledGlyph` subclasses do in fact add fill properties. If you mean actually adding the properties on `FilledGlyph` that's not impossible, but would require additional work to the documentation automation to make sure that the help strings for the inherited properties can still be customized on a per-class basis. 

Do you then imagine this is what things  would look like for users?
```
attrs:
    FilledGlyph:
        fill_alpha: 0.5
```

**Comment by jeremy9959** [comment]
My thought was yes, to use abstract classes and yes, that is what it would look like to the user.
I will do some experimentation to see what this is like in practice.  Incidentally I also understand
what is going on with #6821.  I can make a PR to clear that up but before I do I could use some guidance as to what seems most reasonable, it's described there.
